### PR TITLE
[EDU-765] Update jwt authentication tutorial

### PR DESCRIPTION
**Pull request by fliptopbox**
## Description

As part of the stocktaking of all tutorials (see: https://ably.atlassian.net/browse/EDU-765) 
This change repated to the updated tutorial PR  https://github.com/ably/tutorials/pull/104
 we found inconsistencies, and failing code samples, for the most visited tutorial. 

This PR is a QUICK-FIX the following:

- the Textile instructions do not match the Github repo code snippet (see: https://github.com/ably/tutorials/tree/jwt-authentication-nodejs-fix) that fixed this mismatch
- the input form on the demo is not used to set the `clientId` and is redundant, the fix corrects this redundancy
- the JavaScript ES6 updates and styling changes for: use of const, and spread operator, and fat arrow functions.

# Related work:
- Update the live demo CSS for docs on `Website` page [Website PR](https://github.com/ably/website/pull/4736)
- Update the `docs` tutorial, to use the update screenshot and updated JavaScript  [docs tutorial PR](https://github.com/ably/tutorials/pull/104)
- Update the tutorials `source code`, and replace out-of-date ably logo. [Source code PR](https://github.com/ably/docs/pull/1502)

### CODE REVIEW

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:178]:
```suggestion
The Ably platform expects you to send the time-to-live (@expiresIn@) and @keyId@ as part of the @jwtOptions@ parameter. You can optionally configure capabilities for the client using this token or a client ID, these will be part of the payload. Add the following code:
```

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:180]:
```suggestion
7. In this step you will create a simple web server using the "Express web framework":https://www.npmjs.com/package/express to serve static files. In a later step you will also use this server to create an API route for clients to authenticate with.
```

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:192]: User -> user

General note: Don't mix "you" and "we". Use "you" if it's something the user does, "we" if it's something Ably does. E.g. "At Ably, we provide ..."

Suggestion: "In a production application you would check that the user exists and is authorized to use your service before issuing a JWT with their user ID as the @clientId@. In this example you will just use the user ID they provide."

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:194]:
```suggestion
In the @index.html@ file, instantiate the Ably client library with a "@ClientOptions@:/realtime-sdk#client-options object, passing in your authentication endpoint URL as the @authUrl@ property. Append the user's input to this URL as a query string in the format: `/auth?clientId={username}:{password}`. You can access the @clientId@ parameter in the query string using the Express @Request@ object, as @req.query.clientId@
```

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:222]: Suggest putting this _before_ the code sample after the introductory text that says what it does. "... This code is invoked when a client application makes a request to the @/auth@ route on your server."

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:307]: We said this in an earlier step. Probably better to lose this sentence and just remind the reader what they did in that step?

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:315]:
```suggestion
Now you are ready to test your application.
```

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:327]:
```suggestion
4. Visit @http://localhost:3000@ in your browser (because your Express server is listening on port 3000).
```

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:329]: Suggest making the opening of the browser console a step in its own right, otherwise there's not really any point in having the `console.log()` statements.

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:346]:
```suggestion
10. Read about third-party serverless integrations with "Netlify Identity for Ably authentication":https://ably.com/blog/netlify-indentity-jwt
```

**Review by marklewin — CHANGES_REQUESTED** [review]
Looks good. Love the fancy new dialog 👏 I've made a few suggestions to help make the instructions a bit clearer and fix a couple of grammar/typo things.

File: content/tutorials/jwt-authentication.textile
Comment (fliptopbox) [reply content/tutorials/jwt-authentication.textile:174]: thanks. good catch, i refactored that.

File: content/tutorials/jwt-authentication.textile
Comment (fliptopbox) [reply content/tutorials/jwt-authentication.textile:307]: true, true.

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:56]: authenticate

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:62]: This live demo doesn't really show anything useful - it just tells you that you are authenticated*. I think it would be much more valuable to display the JWT and the response here. WDYT?

* Also, I don't think we need the textbox output AND the alert box

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:142]: I think it might be worth breaking down the individual parts of the key here, especially as you do `split()` on it in step 2.7?

```suggestion
3. Copy your API key to the clipboard, from your Ably Dashboard. It will be of the form @<app ID>.<key ID>:<key secret>@. For example: @12AB3a.ab3GOm:c-B4Emryq9aFmOxP@.
```

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:190]: "Middleware" refers to a collection of middleware functions. So "a middleware _function_" is the correct way to refer to one of them.

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:47]: Would be good to link to the Ably JWT docs here: https://ably.com/docs/api/realtime-sdk/authentication#ably-jwt

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:224]:
```suggestion
You have used the "jsonwebtoken":https://www.npmjs.com/package/jsonwebtoken npm library to sign a JWT using the Ably API key which you obtained from your dashboard. The callback function returns an error parameter and the @tokenId@. If there is an error, you log it to the console. If the operation is successful, you send the JWT back to the client. This JWT contains a token that the client can use to authenticate with:

```json
{
       "token": "xVLyHw.Dtxd9tuz....EXAMPLE",
       "capability": "{\"*\":[\"*\"]}"
       "clientId": "client@example.com",
       "expires": 1449745287315,
       "keyName": "xVLyHw.lCE5Tg",
       "issued": 1449741687315,
     } 

```
```

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:295]: Perhaps "Fetching Ably JWT from auth server"?

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:299]: "Client connected to the Ably platform using an Ably JWT"

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:325]:
```suggestion
The server will output its log messages to your terminal.
```

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:331]:
```suggestion
5. Open the browser developer console, to see messages logged by the client.
```

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:333]:
```suggestion
6. Enter any values into the @username@ and @password@ fields and click **login**.
```

**Review by marklewin — CHANGES_REQUESTED** [review]
Great start. A few typos and rewording suggestions and also for improving the live demo.

File: content/tutorials/jwt-authentication.textile
Comment (fliptopbox) [reply content/tutorials/jwt-authentication.textile:299]: Success! You're connected with a JWT token.

File: content/tutorials/jwt-authentication.textile
Comment (fliptopbox) [reply content/tutorials/jwt-authentication.textile:295]: Requesting JWT token from auth server?

File: content/tutorials/jwt-authentication.textile
Comment (fliptopbox) [reply content/tutorials/jwt-authentication.textile:47]: nice!

File: content/tutorials/jwt-authentication.textile
Comment (fliptopbox) [reply content/tutorials/jwt-authentication.textile:62]: true, but on the tutorial page it does server a purpose, that is the user might not have a console open, so clicking login will have no "side-effect". On the localhost demo, i agree 100%. but I will leave it for now.

My preference would be to raise a response message in the UI.

File: content/tutorials/jwt-authentication.textile
Comment (fliptopbox) [reply content/tutorials/jwt-authentication.textile:295]: Requesting ....

**Review by marklewin — APPROVED** [review]
LGTM!

File: content/tutorials/jwt-authentication.textile
Comment (tbedford) [inline content/tutorials/jwt-authentication.textile:18]: Why is the last updated in January!

**Review by tbedford — CHANGES_REQUESTED** [review]
I'm a bit confused. The ticket said there was a mismatch between the code in the [repo](https://github.com/ably/jwt-auth-node-tutorial) and the tutorial, but the code in the repo (which is linked to from the tutorial itself) has [not been updated](https://github.com/ably/jwt-auth-node-tutorial) - so now it doesn't match with the tutorial. If you download the repo code, it won't match the tutorial.

Also, there are some formatting errors, which should be picked up straight away. Tutorials *must* be reviewed using the website view - not just Nanoc.

image

I've not checked all steps in this tutorial, but would like to do so before it is merged. So I've added the `do not merge` label. Thanks.

File: content/tutorials/jwt-authentication.textile
Comment (fliptopbox) [reply content/tutorials/jwt-authentication.textile:18]: my bad.

**Comment by fliptopbox** [comment]
@tbedford the code for the updated repo is here https://github.com/ably/tutorials/tree/jwt-authentication-nodejs-fix
I have added this to the PR description.

**Comment by tbedford** [comment]
> @tbedford the code for the updated repo is here https://github.com/ably/tutorials/tree/jwt-authentication-nodejs-fix
I have added this to the PR description.

Why? I thought we were moving tutorial code out of there - as was done previously? Also, if that is the case, then the links to the repo in the tutorial text are wrong and will need to be updated. Thanks.

**Comment by fliptopbox** [comment]
i will re base that commit into the original. `jwt-authentication` that is just a working branch

**Comment by tbedford** [comment]
> i will re base that commit into the original. jwt-authentication that is just a working branch

No, what I mean is that currently that tutorial has its own repo here:

https://github.com/ably/jwt-auth-node-tutorial

And that is what is currently linked to in the tutorial text. If you're moving the tutorial code back to the tutorials repo, then the links in the tutorial text need to be updated accordingly, because they currently link to the stand alone repo.

**Comment by fliptopbox** [comment]
regarding the move the code out. to where? ATM there is not new destination for tutorial fixes, so i have created a branch to do the work, which will then merge into the original and replace the out-of-date repo. i think that the "new" tutorial repo project is stalled, so in the mean time i presume we should continue working on this repo

**Comment by fliptopbox** [comment]
> > i will re base that commit into the original. jwt-authentication that is just a working branch
> 
> No, what I mean is that currently that tutorial has its own repo here:
> 
> https://github.com/ably/jwt-auth-node-tutorial
> 
> And that is what is currently linked to in the tutorial text. If you're moving the tutorial code back to the tutorials repo, then the links in the tutorial text need to be updated accordingly, because they currently link to the stand alone repo.

I was not intending to move the repo code into tutorials. it will stay where it is.

**Comment by tbedford** [comment]
> I was not intending to move the repo code into tutorials. it will stay where it is.

Currently the tutorial code for the JWT auth tutorial is here:

https://github.com/ably/jwt-auth-node-tutorial

Have you looked at that repo at all?

**Comment by marklewin** [comment]
> > @tbedford the code for the updated repo is here https://github.com/ably/tutorials/tree/jwt-authentication-nodejs-fix
> > I have added this to the PR description.
> 
> Why? I thought we were moving tutorial code out of there - as was done previously? Also, if that is the case, then the links to the repo in the tutorial text are wrong and will need to be updated. Thanks.

As per [this DR](https://ably.atlassian.net/l/cp/B8shL6p2), we're keeping the tutorial solution code in the `tutorials` repo for now. One branch per tutorial, but labelled commits for the individual steps. So the standalone repo needs to go back into `tutorials`.

**Comment by tbedford** [comment]
> As per [this DR](https://ably.atlassian.net/l/cp/B8shL6p2), we're keeping the tutorial solution code in the tutorials repo for now. One branch per tutorial, but labelled commits for the individual steps. So the standalone repo needs to go back into tutorials.

Great, thanks for the clarification.

**Comment by fliptopbox** [comment]
@marklewin and @tbedford I have update the URLs in the textile file to point at the correct branch on our tutorials repo, including the links that reference blobs (like a link to the entire CSS file).
Also I corrected the formatting issues that @tbedford pointed out for the Website view.

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:169]: At the bottom of step 6 there is an "add the following code:" that doesn't show any code 😕

File: content/tutorials/jwt-authentication.textile
Comment (marklewin) [inline content/tutorials/jwt-authentication.textile:325]: Typos:

- "assign the path of *the* auth server"
- "the @/auth@ path of *the* auth server"

**Review by marklewin — CHANGES_REQUESTED** [review]
Checked in website view and looking good - just a couple of minor points.

File: content/tutorials/jwt-authentication.textile
Comment (fliptopbox) [reply content/tutorials/jwt-authentication.textile:169]: :scissors:

**Review by marklewin — APPROVED** [review]
LGTM

File: content/tutorials/jwt-authentication.textile
Comment (tbedford) [inline content/tutorials/jwt-authentication.textile:320]: I'm not sure why this text was changed - it now does not correspond to the original code and tutorial text, which was: `alert("Client successfully connected to Ably using JWT auth")` The new text doesn't really work because JWT is a JSON Web Token, so "Success! You are connected with a JWT token" would be "JSON Web Token token"!

I would stick with the original text as this is what is show in the tutorial text anyway:

image
 
Note, same comment applies to line 319.

File: content/tutorials/jwt-authentication.textile
Comment (tbedford) [inline content/tutorials/jwt-authentication.textile:366]: > https://github.com/ably/jwt-auth-node-tutorial.

Wrong link to tutorial code. Also, minor point, please remove the full stop from the end of the line, as none of the other lines have it.

**Review by tbedford — CHANGES_REQUESTED** [review]
A few minor points, otherwise looks good.

File: content/tutorials/jwt-authentication.textile
Comment (fliptopbox) [reply content/tutorials/jwt-authentication.textile:320]: corrected both instances

File: content/tutorials/jwt-authentication.textile
Comment (fliptopbox) [reply content/tutorials/jwt-authentication.textile:366]: - the link is correct (merged working branch)
- removed the .

Comment (tbedford) [reply content/tutorials/jwt-authentication.textile:366]: No link is not correct. It is currently:

https://github.com/ably/jwt-auth-node-tutorial

It should be:

https://github.com/ably/tutorials/tree/jwt-authentication-nodejs

To align with the other links in the tutorial.

File: content/tutorials/jwt-authentication.textile
Comment (tbedford) [inline content/tutorials/jwt-authentication.textile:355]: > message "1Client successfully connected

You introduced a typo here.

File: content/tutorials/jwt-authentication.textile
Comment (tbedford) [inline content/tutorials/jwt-authentication.textile:366]: Period removed but link not changed.

**Review by tbedford — CHANGES_REQUESTED** [review]

File: content/tutorials/jwt-authentication.textile
Comment (fliptopbox) [reply content/tutorials/jwt-authentication.textile]: sorry, it got my wires crossed. fixing this now.

**Review by tbedford — APPROVED** [review]
I've run through and tested. LGTM!

**Pull request merged**
